File: Projects/DMDJava/fontbox/src/main/java/org/apache/fontbox/cff/CharStringHandler.vb
Namespace org.apache.fontbox.cff

    '/**
    ' * A Handler for CharStringCommands.
    ' * 
    ' * @author Villu Ruusmann
    ' * @version $Revision$
    ' */
    Public MustInherit Class CharStringHandler


        '/**
        ' * Handler for a sequence of CharStringCommands.
        ' * 
        ' * @param sequence of CharStringCommands
        ' * 
        ' * @return may return a command sequence of a subroutine
        ' */
        '@SuppressWarnings(value = { "unchecked" })
        Public Overridable Function handleSequence(ByVal sequence As List(Of Object)) As List(Of NInteger)
            Dim numbers As List(Of NInteger) = Nothing
            Dim offset As Integer = 0
            Dim size As Integer = sequence.size()
            For i As Integer = 0 To size - 1
                Dim [object] As Object = sequence.get(i)
                If (TypeOf ([object]) Is CharStringCommand) Then
                    If (numbers Is Nothing) Then
                        numbers = sequence.subList(offset, i)
                    Else
                        numbers.addAll(sequence.subList(offset, i))
                    End If
                    Dim stack As List(Of NInteger) = handleCommand(numbers, DirectCast([object], CharStringCommand))
                    If (stack IsNot Nothing AndAlso Not stack.isEmpty()) Then
                        numbers = stack
                    Else
                        numbers = Nothing
                    End If
                    offset = i + 1
                End If
            Next
            If (numbers IsNot Nothing AndAlso Not numbers.isEmpty()) Then
                Return numbers
            Else
                Return Nothing
            End If
        End Function

        '/**
        ' * Handler for CharStringCommands.
        ' *  
        ' * @param numbers a list of numbers
        ' * @param command the CharStringCommand
        ' * 
        ' * @return may return a command sequence of a subroutine
        ' */
        Public MustOverride Function handleCommand(ByVal numbers As List(Of NInteger), ByVal command As CharStringCommand) As List(Of NInteger)


    End Class

End Namespace

File: Projects/FinSeAJava/fontbox/src/main/java/org/apache/fontbox/cff/CharStringHandler.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * Me work for additional information regarding copyright ownership.
 * The ASF licenses Me file to You under the Apache License, Version 2.0
 * (the "License"); you may not use Me file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.fontbox.cff;


import java.util.List;

/**
 * A Handler for CharStringCommands.
 * 
 * @author Villu Ruusmann
 * @version $Revision$
 */
public abstract class CharStringHandler
{

    /**
     * Handler for a sequence of CharStringCommands.
     * 
     * @param sequence of CharStringCommands
     * 
     * @return may return a command sequence of a subroutine
     */
    @SuppressWarnings(value = { "unchecked" })
    public List(Of NInteger) handleSequence(List(Of Object) sequence)
    {
        List(Of NInteger) numbers = null;
        int offset = 0;
        int size = sequence.size();
        for (int i = 0; i < size; i++)
        {
            Object object = sequence.get(i);
            if (object instanceof CharStringCommand)
            {
                if (numbers Is Nothing)
                    numbers = (List) sequence.subList(offset, i);
                else 
                    numbers.addAll((List) sequence.subList(offset, i));
                List(Of NInteger) stack = handleCommand(numbers, (CharStringCommand) object);
                if (stack IsNot Nothing && !stack.isEmpty())
                    numbers = stack;
                else
                    numbers = null;
                offset = i + 1;
            }
        }
        if (numbers IsNot Nothing && !numbers.isEmpty())
            return numbers;
        else
            return null;
    }
    /**
     * Handler for CharStringCommands.
     *  
     * @param numbers a list of numbers
     * @param command the CharStringCommand
     * 
     * @return may return a command sequence of a subroutine
     */
    public abstract List(Of NInteger) handleCommand(List(Of NInteger) numbers, CharStringCommand command);
}